HTML API: Add support for missing FRAMESET and "after" insertion modes.

dmsnell opened the pull request:
Trac ticket: Core-61576.

As part of work to add more spec support to the HTML API, this patch
adds support for the FRAMESET-related insertion modes, as well as the
set of missing _after_ insertion modes. These modes run at the end of
parsing a document, closing it and taking care of any lingering tags.

Developed in https://github.com/wordpress/wordpress-develop
Discussed in https://core.trac.wordpress.org/ticket/61576

See #61576.

## `html5lib` tests

```diff
-Tests: 1497, Assertions: 975, Skipped: 522.
+Tests: 1494, Assertions: 1074, Skipped: 420.
```

dmsnell committed:
HTML API: Add support for missing FRAMESET and "after" insertion modes.

As part of work to add more spec support to the HTML API, this patch
adds support for the FRAMESET-related insertion modes, as well as the
set of missing _after_ insertion modes. These modes run at the end of
parsing a document, closing it and taking care of any lingering tags.

Developed in https://github.com/wordpress/wordpress-develop/pull/7165
Discussed in https://core.trac.wordpress.org/ticket/61576

See #61576.

dmsnell commented:
@sirreal there are failures in the tests that look like BODY isn't closing. If you have insight, I'd appreciate it.

sirreal committed:
Fix #funky_comment token name (#funky-comment)

sirreal commented:
> @sirreal there are failures in the tests that look like BODY isn't closing. If you have insight, I'd appreciate it.

I've been reviewing the specification and this may be very tricky! The body tag never seems to close, but things _may be_ inserted outside of it 😵 

When the BODY and HTML tags close, _nothing changes the stack of open elements_. This is different from most other tags and insertion mode transitions. There's no _pop_ here!

> An end tag whose tag name is "body"
> …
> Switch the [insertion mode](https://html.spec.whatwg.org/multipage/parsing.html#insertion-mode) to "[after body](https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-afterbody)".

In "after body" (BODY closed) and "after after body" (HTML closed), comments are inserted in-place (outside BODY or HTML elements) but other things are handled using rules for "in body" insertion mode. The things that are inserted in place have special handling like "Insert a comment as the last child of the first element in the stack of open elements (the html element)" or "Insert a comment as the last child of the Document object."

This is fascinating. We basically operate just like we're in body with a few invisible exceptions that are inserted out of place. The stack of open elements even continues to be modified!

```html
<i><s>
</body>
<!-- "after body" -->
body » i » s » #text
</s><em>
</html>
<!-- "after after body" -->
body » i » em » #text
</i></em>
body » #text
```

![Screenshot 2024-08-09 at 10 33 11](https://github.com/user-attachments/assets/a840dfda-406a-448b-82c4-feda6a798bbc)
![Screenshot 2024-08-09 at 10 33 23](https://github.com/user-attachments/assets/466f2336-0e5b-4257-bafa-90b0bd397d1e)

---

This is similar to TABLE elements, where some elements may be bumped before the table ([_foster parenting_](https://html.spec.whatwg.org/multipage/parsing.html#foster-parent)). Some things, mostly comments, are bumped to after BODY or HTML nodes.

It also has some similarity to FORM element closers that may be removed from the stack of open elements out-of-place.

---

I'm not sure how best to handle this 😕 Reviewing the spec, I believe that only comments are inserted outside of the BODY and HTML nodes. I wonder if we could collect these in-place comments in a couple of lists then iterate through them then the document is finished.

**Update:**

I had overlooked that "after body" and "after after body" actually switch to mode back to body and reprocess the token, so we effectively step out of body or html tags, have an opportunity to insert some comments, then switch back if anything else is found. We do not process the tokens using the rules for "in body".

> Switch the insertion mode to "in body" and reprocess the token.

sirreal committed:
Add presumptuous tag handling to after(after) body

sirreal committed:
Add missing spec quote to after body anything else

sirreal commented:
I explored some ideas for dealing with those comments in https://github.com/dmsnell/wordpress-develop/pull/18.

For now, I'm going to push a change to this PR to switch the bail conditions in after body. It seems better to bail on comments outside of `BODY` and `HTML` nodes (we might even be able to skip this) and support a `</body>` tag in the middle of the document that has almost no impact if text or tags follow it.

sirreal committed:
Support after body elements but not comments

Ideally, we could support all of this and only bail if the processor
prints a comment and then re-enters.

sirreal committed:
Revert "Support after body elements but not comments"

This reverts commit a1304b5b9b77f93294fc832a5a102e8dcea40c3c.

sirreal committed:
Try allowing outside-of-body content

Some content can exist outside of BODY and HTML tags. See the after body
and after after body insertion modes.

These modes are difficult because the tree may be constructed
out-of-order. Disallow this through the use of a few flags that are used
to bail if possible out-of-order behavior is detected.

This allows some HTML to be processed as long as elements are
encountered in-order. Although the body tag may close, elements can
still be found in-order. For example:

`</html>x</body>y</body></!></html></!>` is not problematic because text
nodes are found and return to "in body" insertion mode before any
content is produced outside of the body or html nodes:

Document
  ├HTML
  │ ├── HEAD
  │ ├── /HEAD
  │ ├── BODY
  │ │    ├── #text: "x"
  │ │    └── #text: "y"
  │ ├── /BODY
  │ └── #funky-comment: "!"
  ├── /HTML
  └── #funky-comment: "!"

However, as soon as content is produced outside if body, the processor
will bail if it attempts to return to in body or to produce content
inside body tags. For example `</html></!></body>x` bails. It would
produce this tree:

Document
  ├HTML
  │ ├── HEAD
  │ ├── /HEAD
  │ └── BODY
  │      └── #text: "x"
  ├── /HTML
  └── #funky-comment: "!"

In this case, the `#funky-comment` was found first in the document root,
then a text node "x" was added to the body. These out-of-order nodes are
disallowed and bail.

sirreal committed:
fixup! Try allowing outside-of-body content

sirreal committed:
Handle presumptuous tags everywhere comments are handled

sirreal committed:
Apply same outside-html allowed if in order technique to frameset

sirreal committed:
Fix typo in bail message

sirreal committed:
Fix text handling in frameset insertion modes

These modes effectively ignore non-whitespace (not supported) and insert
whitespace text nodes under HTML node.

Bail if out-of-order behavior is detected.

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -4001,7 +4007,85 @@ private function step_in_template(): bool {
+			/*
+			 * > A comment token
+			 */
+			case '#cdata-section':
```
note to self: I'm not sure how these got in here, but they aren't comments, and I don't see a line in the spec for them. I'm thinking it could have been a copy/paste error I missed.

dmsnell commented:
> I'm not sure how best to handle this 😕 Reviewing the spec, I believe that only comments are inserted outside of the BODY and HTML nodes. I wonder if we could collect these in-place comments in a couple of lists then iterate through them then the document is finished.

This seems like a plausible idea, @sirreal. Presumably we could reuse the event queue, such that we track these comments in those lists, and `next_token()` would pull them out when it sees the POP event for BODY and then for HTML.

From a performance standpoint, I don't see this being that troublesome, especially if all we stored in those lists were the `WP_HTML_Token` or bookmark name.

sirreal commented:
> > I wonder if we could collect these in-place comments in a couple of lists then iterate through them then the document is finished.
> 
> This seems like a plausible idea, @sirreal. Presumably we could reuse the event queue, such that we track these comments in those lists, and `next_token()` would pull them out when it sees the POP event for BODY and then for HTML.

I implemented that in this PR in a1304b5b9b77f93294fc832a5a102e8dcea40c3c. I reverted the change. I had issues getting the text for the comments outside of body because `get_modifiable_text` is a method on the Tag Processor and it expects to operate on the token it's parsing. It may be enough to seek to all the comments to get their text, I did not try that.

sirreal committed:
Revert "smart" out-of-body out-of-html content handling

Simplify this PR to handle common scenarios and avoid deviation from specification.

sirreal committed:
Merge branch 'trunk' into html-api/support-after-insertion-modes

sirreal reviewed (approved):
This seems to be in a good place 👍  We get a good increase in HTML5lib-tests which also helps my confidence:

```diff
 OK, but incomplete, skipped, or risky tests!
-Tests: 1498, Assertions: 930, Skipped: 568.
+Tests: 1495, Assertions: 1026, Skipped: 469.
```

sirreal committed:
Remove cdata section handling from new insertion modes

These should not have appeared as CDATA cannot appear in HTML.

dmsnell committed:
Merge remote-tracking branch 'upstream/trunk' into html-api/support-after-insertion-modes

dmsnell commented:
Merged in [[58926]](https://core.trac.wordpress.org/changeset/58926)
https://github.com/wordpress/wordpress-develop/commit/15dca4effa8f1786176aa7acd6f3076c48f4a1fa
